File: day07/solution.py
"""Day 7: Laboratories - simulating tachyon beams through a splitter manifold."""

from pathlib import Path


def simulate_beams(grid: list[str]) -> int:
    """Simulate beams traveling down through the manifold, count splits."""
    # find where the beam enters (the 'S')
    start_col = grid[0].index("S")

    # track active beam positions as a set - handles merging automatically
    beams = {start_col}
    total_splits = 0

    # simulate each row from top to bottom
    for row in range(1, len(grid)):
        line = grid[row]
        new_beams = set()

        for col in beams:
            # check if this beam hits a splitter
            if col < len(line) and line[col] == "^":
                # beam hit a splitter! count it and spawn left/right
                total_splits += 1
                if col - 1 >= 0:
                    new_beams.add(col - 1)
                if col + 1 < len(line):
                    new_beams.add(col + 1)
            else:
                # empty space - beam continues straight down
                if 0 <= col < len(line):
                    new_beams.add(col)

        beams = new_beams

        # if no beams left, we're done
        if not beams:
            break

    return total_splits


def count_timelines(grid: list[str]) -> int:
    """Count distinct timelines using many-worlds interpretation."""
    start_col = grid[0].index("S")

    # now we track particle COUNTS at each position, not just presence
    # particles at same position are still distinct timelines
    particles = {start_col: 1}

    for row in range(1, len(grid)):
        line = grid[row]
        new_particles: dict[int, int] = {}

        for col, count in particles.items():
            if col < len(line) and line[col] == "^":
                # particle splits into two timelines (left and right)
                if col - 1 >= 0:
                    new_particles[col - 1] = new_particles.get(col - 1, 0) + count
                if col + 1 < len(line):
                    new_particles[col + 1] = new_particles.get(col + 1, 0) + count
            else:
                # continues down - preserves all timeline counts
                if 0 <= col < len(line):
                    new_particles[col] = new_particles.get(col, 0) + count

        particles = new_particles
        if not particles:
            break

    # total timelines = sum of all particle counts
    return sum(particles.values())


def main():
    text = (Path(__file__).parent / "input.txt").read_text()
    grid = text.strip().split("\n")

    part1 = simulate_beams(grid)
    print(f"Part 1: {part1}")

    part2 = count_timelines(grid)
    print(f"Part 2: {part2}")


if __name__ == "__main__":
    main()

File: day07/solution.go
// Advent of Code 2025 - Day 7: Laboratories (tachyon beam simulation)
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// simulateBeams counts how many times beams are split (Part 1)
// Beams at the same position merge - we only care about unique positions
func simulateBeams(grid []string) int {
	// find where the beam enters (the 'S')
	startCol := strings.Index(grid[0], "S")

	// Go doesn't have a built-in set type, so we use map[int]struct{}
	// struct{} takes zero bytes - it's the idiomatic "set" pattern
	beams := map[int]struct{}{startCol: {}}
	totalSplits := 0

	// simulate each row from top to bottom
	for row := 1; row < len(grid); row++ {
		line := grid[row]
		newBeams := make(map[int]struct{})

		for col := range beams {
			// check if this beam hits a splitter
			if col < len(line) && line[col] == '^' {
				// beam hit a splitter! count it and spawn left/right
				totalSplits++
				if col-1 >= 0 {
					newBeams[col-1] = struct{}{}
				}
				if col+1 < len(line) {
					newBeams[col+1] = struct{}{}
				}
			} else {
				// empty space or out of bounds - beam continues down
				if col >= 0 && col < len(line) {
					newBeams[col] = struct{}{}
				}
			}
		}

		beams = newBeams

		// if no beams left, we're done early
		if len(beams) == 0 {
			break
		}
	}

	return totalSplits
}

// countTimelines counts total timelines using many-worlds interpretation (Part 2)
// Particles at the same position are STILL distinct - they don't merge
func countTimelines(grid []string) int64 {
	startCol := strings.Index(grid[0], "S")

	// now we track particle COUNTS at each position
	// map[column]count - multiple particles can stack at same column
	particles := map[int]int64{startCol: 1}

	for row := 1; row < len(grid); row++ {
		line := grid[row]
		newParticles := make(map[int]int64)

		for col, count := range particles {
			if col < len(line) && line[col] == '^' {
				// particle splits into two timelines (left and right)
				// each timeline inherits all the counts from this position
				if col-1 >= 0 {
					newParticles[col-1] += count
				}
				if col+1 < len(line) {
					newParticles[col+1] += count
				}
			} else {
				// continues down - preserves all timeline counts
				if col >= 0 && col < len(line) {
					newParticles[col] += count
				}
			}
		}

		particles = newParticles
		if len(particles) == 0 {
			break
		}
	}

	// total timelines = sum of all particle counts at final positions
	var total int64 = 0
	for _, count := range particles {
		total += count
	}
	return total
}

func loadInput(filename string) []string {
	file, _ := os.Open(filename)
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	// trim trailing empty lines
	for len(lines) > 0 && strings.TrimSpace(lines[len(lines)-1]) == "" {
		lines = lines[:len(lines)-1]
	}

	return lines
}

func main() {
	// example from puzzle description
	example := []string{
		".......S.......",
		"...............",
		".......^.......",
		"...............",
		"......^.^......",
		"...............",
		".....^.^.^.....",
		"...............",
		"....^.^...^....",
		"...............",
		"...^.^...^.^...",
		"...............",
		"..^...^.....^..",
		"...............",
		".^.^.^.^.^...^.",
		"...............",
	}

	fmt.Println("--- Part 1 ---")
	fmt.Printf("Example: %d (expected 21)\n", simulateBeams(example))

	input := loadInput("day07/input.txt")
	fmt.Printf("Answer:  %d\n", simulateBeams(input))

	fmt.Println("\n--- Part 2 ---")
	fmt.Printf("Example: %d (expected 40)\n", countTimelines(example))
	fmt.Printf("Answer:  %d\n", countTimelines(input))
}
